src/control/server/engine/exec.go

```diff
@@ -126,19 +145,16 @@ func (r *Runner) Signal(signal os.Signal) error {
 		return nil
 	}
 
-	r.log.Debugf("Signalling I/O Engine instance %d (%s)", r.Config.Index, signal)
-
-	return r.cmd.Process.Signal(signal)
+	r.sigCh <- signal
+	return <-r.sigErrCh
 }
 
-// GetLastPid returns the PID after runner has exited, return
-// zero if no cmd or ProcessState exists.
+// GetLastPid returns the PID after runner has exited. It
+// is usually an error to call this with a running process,
+// as the PID is not available on the channel until the
+// process exits, and therefore the call may block.
 func (r *Runner) GetLastPid() uint64 {
-	if r.cmd == nil || r.cmd.ProcessState == nil {
-		return 0
-	}
-
-	return uint64(r.cmd.ProcessState.Pid())
+	return uint64(<-r.pidCh)
```

Comment:
I would be more comfortable this was nonblocking:

```suggestion
	if r.IsRunning() {
	        return 0
	}
	return uint64(<-r.pidCh)
```

Reply:
In that case the function isn't really returning the last pid, is it? This `GetLastPid()` method is only called in `EngineInstance.exit()`, so it doesn't really make sense to me that it could be called with a running process.